### PR TITLE
MGDOBR-1093: Processor is not updating to Ready state from Failed state

### PR DESCRIPTION
**Pull request by manstis**
[https://issues.redhat.com/browse/MGDOBR-1093](https://issues.redhat.com/browse/MGDOBR-1093)

See JIRA for details.

## PR owner checklist

Please make sure that your PR meets the following requirements:

- [x] Your code is properly formatted according to [this configuration](https://github.com/kiegroup/kogito-runtimes/tree/main/kogito-build/kogito-ide-config).  
  *Run `mvn clean verify -DskipTests` so that imports are correctly set.*
- [x] Your commit messages are clear and reference the JIRA issue e.g: "[MGDOBR-1] - $clear_explanation_of_what_you_did"
- [x] The layers in the `kustomize` folder have been updated accordingly.
- [x] All new functionality is tested
- [x] Pull Request title is properly formatted: `MGDOBR-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

## PR reviewer(s) checklist

- [ ] Check if code and Github action workflows have been modified and if the modification is safe
- [ ] If the modification is safe, add the `safe to test` label

<details>
<summary>
How to trigger pipelines and use the bots:
</summary>

- <b>Run the end to end pipeline</b>  
  Annotate the pull request with the label: `safe to test`. If you want to run the pipeline again, remove and add it again.

- <b>Rebase the pull request</b>  
  Comment with: `/rebase`.
- <b>Deploy BOT</b>
  Comment with `/deploy <dev,stable> when the PR has been merged to deploy to a target environment.

</details>

### CODE REVIEW

**Comment by Aakanksha0311** [comment]
@manstis I see that it is failing to update processor from failed state to ready state for the slack source processor scenario.

**Comment by manstis** [comment]
> I see that it is failing to update processor from failed state to ready state for the slack source processor scenario.

@Aakanksha0311 can you provide an example definition? All tests are passing.

**Review by Aakanksha0311 — APPROVED** [review]
LGTM!

**Comment by Srihari1192** [comment]
@manstis Seems this Failing scenario is not handled for error handling processor update

**Comment by Srihari1192** [comment]
@manstis Also I think its good to have unit test coverage from Processor **Failed** status to **Ready** state

**Comment by Srihari1192** [comment]
> @Srihari1192 I have fixed the reported Error Handler scenario and provided a unit test as requested.

@manstis Sure will take a look

**Review by Srihari1192 — APPROVED** [review]
LGTM . Thanks @manstis

File: manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorServiceImpl.java [hunk 1]
```diff
@@ -361,6 +362,19 @@ public Processor updateProcessorStatus(ProcessorManagedResourceStatusUpdateDTO u
                 break;
 
             case UPDATE:
+                // An UPDATE can follow a FAILED_UPDATE when updating a Processor.
+                // If the User has updated a Processor that was previously failed by k8s it has been observed
+                // that the reconciliation loop can first emit an update with the existing FAILED state. This
+                // can lead to the Bridge being FAILED by the BridgeWorker if the Processor is an Error Handler.
+                // Therefore, if a subsequent READY state is emitted we may need to reflect this in the Bridge.
+                if (processor.getType() == ProcessorType.ERROR_HANDLER) {
```
Comment (lucamolteni) [inline manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorServiceImpl.java:370]: Why we want to support this only for error handler processors? Wouldn't this hold for "normal" processors as well?

File: manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/BridgeWorker.java [hunk 2]
```diff
@@ -95,9 +101,7 @@ public Bridge createDependencies(Work work, Bridge bridge) {
                 // This ensures Users cannot update the Bridge Error Handler until it is available.
                 bridge.setDependencyStatus(ManagedResourceStatus.READY);
             } else if (errorHandler.getStatus() == ManagedResourceStatus.FAILED) {
-                bridge.setStatus(ManagedResourceStatus.FAILED);
-                bridge.setDependencyStatus(ManagedResourceStatus.FAILED);
-                propagateProcessorError(bridge);
+                return persistAndPropagateErrorHandlerFailure(bridge, errorHandler);
```
Comment (lucamolteni) [inline manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/BridgeWorker.java:104]: nitpick: I'd do two different methods, one for persisting and the other to propagate the error as they seem unrelated

File: manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/BridgeWorker.java [hunk 3]
```diff
@@ -210,15 +212,15 @@ protected boolean isDeprovisioningComplete(Bridge managedResource) {
...
-    protected void propagateProcessorError(Bridge bridge) {
-        String bridgeId = bridge.getId();
-        String customerId = bridge.getCustomerId();
-        processorService.getErrorHandler(bridgeId, customerId)
-                .ifPresent(errorHandler -> {
-                    bridge.setErrorId(errorHandler.getErrorId());
-                    bridge.setErrorUUID(errorHandler.getErrorUUID());
-                    persist(bridge);
-                });
+    protected Bridge persistAndPropagateErrorHandlerFailure(Bridge bridge, Processor errorHandler) {
```
Comment (lucamolteni) [inline manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/BridgeWorker.java:215]: Again, not sure how these two parts are related

**Review by lucamolteni — APPROVED** [review]
Just a few stylistical comments

File: manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorServiceImpl.java (hunk 1 above)
Comment (manstis) [reply manager/src/main/java/com/redhat/service/smartevents/manager/ProcessorServiceImpl.java:370]: We only propagate Error Handler `Processor` failures to the `Bridge` (otherwise, since they are _internal_, the User is blissfully unaware that something went wrong). If Error Handler `Processor`'s were visible - i.e. listed as a `Processor` in our User facing API - we would not need to propagate the failure.

File: manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/BridgeWorker.java (hunk 2 above)
Comment (manstis) [reply manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/BridgeWorker.java:104]: I do the same thing (persist and propagate) in two different places. 

The single method was added to remove duplicate code.

File: manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/BridgeWorker.java (hunk 3 above)
Comment (manstis) [reply manager/src/main/java/com/redhat/service/smartevents/manager/workers/resources/BridgeWorker.java:215]: It's an internal implementation detail.

I need to persist the `Bridge` object before invoking `BridgeSerrvice.updateBridgeStatus(..)` since this loads the `Bridge` based on ID. I could refactor things to use an in-memory object but the `BridgeService` is a public service and I'd prefer not to pollute its interface to avoid this method.
